### DISCUSSION TITLE
Best way to structure a navigation tree for the frontend?

### DISCUSSION THREAD

**Discussion by rowild** [directus/directus#14137] [Archive]
Is there a recommended way to build a navigation tree for the frontend? I am asking particularly for a navigation structure for SPAs/PWAs, where router come into play. I am searching for a solution that eventually delivers JSON, much in the form of:

```json
{
  "mainNav": [
    {
      "title": "Home",
      "path": "/home"
    },
    {
      "title": "About",
      "path": "about",
      "children": [
        {
          "title": "CV",
          "path": "/about/cv"
        },
        {
          "title": "Musicians",
          "path": "/about/musicians"
        },
        {
          "title": "Recordings",
          "path": "/about/recordings"
        }
      ]
    },
    {
      "title": "Compositions",
      "path": "compositions",
      "children": [
        {
          "title": "All",
          "path": "/compositions?cat=all"
        },
        {
          "title": "Solo",
          "path": "/compositions?cat=solo"
        },
        {
          "title": "Chamber",
          "path": "/compositions?cat=chamber"
        },
        {
          "title": "Ensemble",
          "path": "/compositions?cat=ensemble"
        },
        {
          "title": "(With) Voice",
          "path": "/compositions?cat=voice"
        },
        {
          "title": "Orchestra",
          "path": "/compositions?cat=orchestra"
        },
        {
          "title": "(With) Electronics",
          "path": "/compositions?cat=electronics"
        }
      ]
    },
    {
      "title": "Writings",
      "path": "writings",
      "children": [
        {
          "title": "Writings from John Van der Slice",
          "path": "/texts/from_jvds"
        },
        {
          "title": "Writings about John van der Slice",
          "path": "/texts/about_jvds"
        }
      ]
    },
    {
      "title": "Gallery",
      "path": "galleries",
      "children": [
        {
          "title": "Scores Excerpts",
          "path": "/galleries/scores_excerpts"
        }
      ]
    }
  ],
  "navMetaItems": [
    {
      "title": "Welcome!",
      "path": "/",
      "icon": "fas fa-home fa-2x",
      "isImportant": true
    },
    {
      "title": "Copyright",
      "path": "/copyright",
      "icon": "fas fa-copyright"
    },
    {
      "title": "Disclaimer",
      "path": "/disclaimer",
      "icon": "fas fa-info-circle"
    },
    {
      "title": "Address",
      "path": "/address",
      "icon": "fas fa-envelope-open"
    },
    {
      "title": "Socials",
      "path": "https://www.facebook.com/john.vanderslice.3",
      "external": true,
      "icon": "fab fa-facebook-f"
    }
  ]
}
```

I tried various approaches, but get stuck:
- a Repeater requires to redefine fields menu_item fields for each submenu level
- the concept of the treeview interface seems more like it, but it is not possible to show e.g. the title field of a selected item, which eventually is a problem for an editor

Would you mind to share how you approach this topic? 
Thank you!

**Comment by rowild** [comment]
Anybody? Or is nobody managing navigation with Directus?

**Comment by kamil-malinski** [comment]
I did this with a repeater

beside of the title, the url, the icon etc. one has to enter a tree-path for each entry that defines the path to that element, e.g.

man.entry1
main.entry1.child1
main.entry1.child2
main.entry1.child2.subchild1

etc.

I'm not very happy with that solution but I didn't find a better way to do this with Directus

**Reply by rowild** [reply]
@kamil-malinski Thank you very much for your feedback! – Yes, I tried repeater as well, but I had to recreate all input fields for every child level again. Also, the display in the backend is not really configurable (I would like to define "title" to be displayed in the layouts).

ATM I am fooling around with treeview, but so far ... not really good...

Anybody else having a different approach?
@benhaynes What do the creators of Directus recomment?

**Reply by kamil-malinski** [reply]
> Yes, I tried repeater as well, but I had to recreate all input fields for every child level again

what do you mean by that? the repeater is basically a tool to define a few fields and then create lists of sets with those fields... what is missing is a way to define a tree-hierarchy but I solved that with the tree-path-field like descriped above

**Reply by rowild** [reply]
I think I do not understand what is and how to implement the "tree-path".
What i do is create a repeater field on each menu level again and there – again – create fields for "title","slug" etc
Would you mind telling me what you mean by the "tree-field"? MAybe sharing a screenshot?

**Reply by kamil-malinski** [reply]
I created a collection with an "alias"-field (text) and repeater with the following fields:

treepath (text)
title (text)
url (text)

this allows me to create a navigation e.g. with the alias "main" and a list like that:

**treepath | title | url**
entry1 | Title 1 | /title1
entry2 | Title 2 | /title2
entry1.sub1 | Sub 1 | /title1/sub1
entry1.sub2 | Sub 2 | /title1/sub2
entry1.sub2.category1 | Category 1 | /title1/sub2/category1
entry1.sub2.category2 | Category 2 | /title1/sub2/category2

This generates a json that I convert (within the frontend) to a menu with two main-entries and two levels of sub-entries:

- Title 1
- Title 2
    - Sub 1
    -   Sub 2
        -     Category 1
        -     Category 2

**Reply by rowild** [reply]
Interesting! Thanks for sharing!

But the result of the repeater is flat, right? I get a result like this:

```json
  {
    "data": [
      {
      "id": 1,
      "status": "published",
      "sort": null,
      "user_created": "f657497c-2737-4836-bf01-19a41e243831",
      "date_created": "2022-06-28T09:20:36.362Z",
      "alias": "Menu 1",
      "repeater": [
        {
          "treepath": "item1",
          "title": "Item 1",
          "slug": "/item1"
        },
        {
          "treepath": "item2",
          "title": "Item 2",
          "slug": "item2"
        },
        {
          "treepath": "Item3",
          "title": "Item 3 - with Children",
          "slug": "item3"
        },
        {
          "treepath": "item3.child1",
          "title": "Item 3 - Child 1",
          "slug": "item3/child1"
        },
        {
          "treepath": "item3.child2",
          "title": "Item 3 - Child 2",
          "slug": "item3/child2"
        },
        {
          "treepath": "item3.child3",
          "title": "Item 3 - Child 3",
          "slug": "item3/child3"
        },
        {
          "treepath": "item4",
          "title": "Item 4",
          "slug": "item4"
        },
        {
          "treepath": "item4.child1",
          "title": "Item 4 - Child 1",
          "slug": "item4/child1"
        }
      ]
    }
  ]
}
```

**Reply by kamil-malinski** [reply]
yes, the result of the repeater is flat... I use a php-function to create a nested array with the menu entries from this flat list

**Reply by rowild** [reply]
But if I wanted to achieve nested menues like this:

```json
[...]
"repeater": [
        {
          "treepath": "item1",
          "title": "Item 1",
          "slug": "/item1",
          children: [
            {
              [...]
            }
          ]
        },
 ```
 each level would need another repeater field defined – with again the same input fields... Or am I doing something wrong?

**Reply by rowild** [reply]
I read your answer after my second post - yes, thank you again for sharing! Much appreciated! :-)

**Reply by kamil-malinski** [reply]
yes, but I don't think that's a good approach

**Reply by kamil-malinski** [reply]
> I read your answer after my second post - yes, thank you again for sharing! Much appreciated! :-)

no problem, you are welcome
